pkg/controllers/controlplanemachineset/controller.go

```diff
@@ -240,11 +242,61 @@ func (r *ControlPlaneMachineSetReconciler) ensureFinalizer(ctx context.Context,
 // ensureOwnerReferences determines if any of the Machines within the machineInfos require a new controller owner
 // reference to be added, and then uses PartialObjectMetadata to ensure that the owner reference is added.
 func (r *ControlPlaneMachineSetReconciler) ensureOwnerReferences(ctx context.Context, logger logr.Logger, cpms *machinev1.ControlPlaneMachineSet, machineInfos map[int32][]machineproviders.MachineInfo) error {
-	// TODO: Iterate over the MachineInfos, for each Machine, check the owner references for an owner reference matching
-	// that of the current CPMS (it should be the controller so should be easy to find).
-	// If required, look up the GVK using the rest mapper from the GVR, then uses metav1.PartialObjectMetadata to Patch
-	// the owner references. This should mean we can update the metadata of any type given we know the GVR and existing
-	// ObjectMeta.
+	for _, machineInfo := range machineInfos {
+		for _, mInfo := range machineInfo {
+			if mInfo.MachineRef == nil {
+				continue
+			}
+
+			mObjectMeta := mInfo.MachineRef.ObjectMeta
+			mLogger := logger.WithValues("machineNamespace", mObjectMeta.GetNamespace(), "machineName", mObjectMeta.GetName())
+
+			machineGVK, err := r.RESTMapper.KindFor(mInfo.MachineRef.GroupVersionResource)
+			if err != nil {
+				return fmt.Errorf("error getting GVK for machine: %w", err)
+			}
+
+			machine := &metav1.PartialObjectMetadata{}
+			machine.SetGroupVersionKind(machineGVK)
+			machine.ObjectMeta = mObjectMeta
+
+			if isOwnedByCurrentCPMS(cpms, machine) {
+				mLogger.V(4).Info("Owner reference already present on machine")
+
+				continue
+			}
+
+			patchBase := client.MergeFrom(machine.DeepCopy())
+
+			err = controllerutil.SetControllerReference(cpms, machine, r.Scheme)
+			if err != nil {
```

Comment:
Scope the error to the if block to avoid shadowing
```suggestion
			if err: = controllerutil.SetControllerReference(cpms, machine, r.Scheme); err != nil {
```